### ISSUE TITLE
Frequent Kafka Rebalance and Lag After Upgrading to ClickHouse from 23.8 to 24.3

### ISSUE THREAD

**Issue by bun4uk** [ClickHouse/ClickHouse#71064]
ClickHouse version: 24.3 (upgraded from 23.8)
Cluster setup: 4 nodes (2 shards, 2 replicas)
Data ingestion setup: Kafka Engine -> Materialized View -> MergeTree

Problem Description

After upgrading from ClickHouse 23.8 to 24.3, Kafka consumers frequently enter a rebalance state, causing significant lag during data ingestion. The issue manifests with the following warning in the logs:
```
2024.10.25 13:37:06.109369 [ 2062280 ] {} <Warning> StorageKafka (feedbacks): [rdk:SESSTMOUT] [thrd:main]: Consumer group session timed out (in join-state steady) after 14367 ms without a successful response from the group coordinator (broker 1, last error was Broker: Group rebalance in progress): revoking assignment and rejoining group
```

**Steps Taken:**
- Increased kafka_max_wait_ms to attempt to address the lag and rebalancing issues, but this did not resolve the problem.

**Expected Behavior:**
Kafka should maintain a stable consumer group assignment without frequent rebalancing, as was the case with ClickHouse 23.8.

**Actual Behavior:**
Post-upgrade, Kafka consistently times out and rebalances, causing ingestion lags.

**Request for Assistance:**
Looking for guidance on how to stabilize Kafka ingestion after upgrading to ClickHouse 24.3. Are there any new configurations or adjustments required for Kafka consumers in this version?

**Comment by aol-59** [comment]
Hi @bun4uk , does the issue still persist, or have you found a solution?

**Comment by rchadin** [comment]
I increased the value of the background_message_broker_schedule_pool_size parameter to 64

**Comment by bun4uk** [comment]
> Hi @bun4uk , does the issue still persist, or have you found a solution?
Hello! Unfortunately no.

>I increased the value of the background_message_broker_schedule_pool_size parameter to 64
@rchadin did it help you? I increased it to 200, but I still have rebalances all the time.

**Comment by rchadin** [comment]
Yes, it helped. We adjusted this value, and the number of rebalances almost stopped. We looked at the pool usage by the metric.

**Comment by rchadin** [comment]
You can see consumers in the system.kafka_consumers table. And then draw conclusions. We decided to do this

**Comment by romankydybets** [comment]
After rolling the clickhouse replicas I also see rebalancing happens. 
we have 96 partitions to 96 consumers, it it always jumps. I don't really understand why this happens, look like clickhouse try to open more connections

**Comment by schillerr** [comment]
Check the value of background_message_broker_schedule_pool_size.
For example, like this:

```
SELECT
    (
        SELECT value
        FROM system.metrics
        WHERE metric = 'BackgroundMessageBrokerSchedulePoolTask'
    ) AS tasks,
    (
        SELECT value
        FROM system.metrics
        WHERE metric = 'BackgroundMessageBrokerSchedulePoolSize'
    ) AS pool_size,
    pool_size - tasks AS free_threads
```

In my case, after updating from version **23.8** to version **24.3**, the default value of `background_message_broker_schedule_pool_size`, which equals 16, started being used, even though the `users.xml` file had the following value:

`<background_message_broker_schedule_pool_size>200</background_message_broker_schedule_pool_size>`


To resolve this issue, I updated the configuration by adding the following to the `config.xml` file:

`<background_message_broker_schedule_pool_size>200</background_message_broker_schedule_pool_size>`

This resolved the timeout issues.
